### PR TITLE
[Partitioner] Reuse partition to check whether nodes exist

### PR DESCRIPTION
**Pull request by ghost**
The time complexity of find node whether in NodeList is O(n). Reuse partition to speed up due to partition.nodes is hash table and has same elements.

### CODE REVIEW

**Comment by ezyang** [comment]
@SherlockNoMad if you don't have time to review this please kick it to me, I think I definitely want this in

**Comment by lingzhi98** [comment]
And now there only exist bc_linter error. I am not sure whether this api change can be approved. Can anyone give some advice?

File: torch/fx/passes/infra/partitioner.py [hunk 1]
```diff
@@ -268,7 +268,7 @@ def fuse_partitions(self, partitions: List[Partition], prefix: str = "fused_") -
         # fuse_by_partitions expects partitions in List[List[Node]]: [ [node0, node1], [node2, node3] ]
```
Comment (ezyang) [inline torch/fx/passes/infra/partitioner.py:268]: This comment needs updating I guess?

**Review by ezyang — DISMISSED** [review]

File: torch/fx/passes/infra/partitioner.py (hunk 1 above)
Comment (lingzhi98) [reply torch/fx/passes/infra/partitioner.py:268]: It should be updated. I will submit other PR due to this PR has been merged.

Comment (lingzhi98) [reply torch/fx/passes/infra/partitioner.py:268]: See https://github.com/pytorch/pytorch/pull/135662.

**Comment by izaitsevfb** [comment]
@pytorchbot revert -m "BC-breaking, breaks executorch and internal meta builds" -c ghfirst


`TypeError: fuse_as_graphmodule() missing 1 required positional argument: 'module_name'`

one example of the usage: 
https://github.com/pytorch/executorch/blob/main/exir/lowered_backend_module.py#L744

is it possible to make `partition` the last argument and add a default value to it?

**Comment by lingzhi98** [comment]
> @pytorchbot revert -m "BC-breaking, breaks executorch and internal meta builds" -c ghfirst
> 
> `TypeError: fuse_as_graphmodule() missing 1 required positional argument: 'module_name'`
> 
> one example of the usage: https://github.com/pytorch/executorch/blob/main/exir/lowered_backend_module.py#L744
> 
> is it possible to make `partition` the last argument and add a default value to it?

Has modified this.

File: test/allowlist_for_publicAPI.json [hunk 2]
```diff
@@ -2092,6 +2092,7 @@
     "Node",
     "NodeList",
     "NodeSet",
+    "Optional",
```
Comment (albanD) [inline test/allowlist_for_publicAPI.json:2095]: We do not allow new entries in this file. Would you please be able to cleanup the public API for this file such that nothing is confused as public API.

In this case, your simplest way forward is most likely to define `__all__` there to mark what is actually public there.

**Review by albanD — CHANGES_REQUESTED** [review]

File: test/allowlist_for_publicAPI.json (hunk 2 above)
Comment (lingzhi98) [reply test/allowlist_for_publicAPI.json]: I read the definition of what public api is and think it is suitable to mark Optional as unpublic due to it is the built-in type of python. See this line https://github.com/pytorch/pytorch/pull/135317/files#diff-9d94acf13fbaa041fa5d29da1f79d42d2a361f389706bcc6da67c30c6ec92eacR4

**Comment by lingzhi98** [comment]
@izaitsevfb any comments? Thanks!

**Comment by ezyang** [comment]
need to fix downstream uses. I might just redo this diff from hand in fbcode so I can get the other call sites

**Comment by ezyang** [comment]
Actually I rereviewed and this looks fine

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**